parley/src/layout/data.rs

```diff
@@ -548,4 +557,418 @@ impl<B: Brush> LayoutData<B> {
+                cluster_advance,
+                total_glyphs,
+                cluster_type,
+                inline_glyph_id,
+            );
+        }
+    }
+
+    run_advance
+}
+
+enum ClusterType {
+    LigatureStart,
+    LigatureComponent,
+    Regular,
+    Newline,
+}
+
+impl From<&ClusterType> for u16 {
+    fn from(cluster_type: &ClusterType) -> Self {
+        match cluster_type {
+            ClusterType::LigatureStart => ClusterData::LIGATURE_START,
+            ClusterType::LigatureComponent => ClusterData::LIGATURE_COMPONENT,
+            ClusterType::Regular | ClusterType::Newline => 0, // No special flags
+        }
+    }
+}
+
+fn push_cluster(
+    clusters: &mut Vec<ClusterData>,
+    char_info: &(swash::text::cluster::CharInfo, u16),
+    text_len: u8,
+    cluster_start_char: (usize, char),
+    glyph_offset: u32,
+    advance: f32,
+    total_glyphs: u32,
+    cluster_type: ClusterType,
+    inline_glyph_id: Option<u32>,
+) {
+    let glyph_len = (total_glyphs - glyph_offset) as u8;
+
+    let (final_glyph_len, final_glyph_offset, final_advance) = match cluster_type {
+        ClusterType::LigatureComponent => {
+            // Ligature components have no glyphs, only advance.
+            debug_assert_eq!(glyph_len, 0);
+            (0_u8, 0_u32, advance)
+        }
+        ClusterType::Newline => {
+            // Newline clusters are stripped of their glyph contribution.
+            debug_assert_eq!(glyph_len, 1);
+            (0_u8, 0_u32, 0.0)
+        }
+        _ if inline_glyph_id.is_some() => {
+            // Inline glyphs are stored inline within `ClusterData`
+            debug_assert_eq!(glyph_len, 0);
+            (0xFF_u8, inline_glyph_id.unwrap(), advance)
+        }
+        ClusterType::Regular | ClusterType::LigatureStart => {
+            // Regular and ligature start clusters maintain their glyphs and advance.
+            debug_assert_ne!(glyph_len, 0);
+            (glyph_len, glyph_offset, advance)
+        }
+    };
+
+    clusters.push(ClusterData {
+        info: ClusterInfo::new(char_info.0.boundary(), cluster_start_char.1),
+        flags: (&cluster_type).into(),
+        style_index: char_info.1,
+        glyph_len: final_glyph_len,
+        text_len,
+        glyph_offset: final_glyph_offset,
+        text_offset: cluster_start_char.0 as u16,
+        advance: final_advance,
+    });
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) struct FontMetrics {
+    ascent: i16,
+    descent: i16,
+    leading: i16,
+
+    units_per_em: u16,
+
+    strikethrough_offset: i16,
+    strikethrough_size: i16,
+
+    underline_offset: i16,
+    underline_size: i16,
+}
+
+impl FontMetrics {
+    fn from(font: &skrifa::FontRef<'_>) -> Self {
+        use skrifa::raw::{TableProvider, tables::os2::SelectionFlags};
+        // NOTE: This _does not_ copy harfrust's metrics behaviour (https://github.com/harfbuzz/harfrust/blob/a38025fb336230b492366740c86021bb406bcd0d/src/hb/glyph_metrics.rs#L55-L60).
```

Comment:
Instead, this copies the behaviour from Swash (choosing typographic metrics only if the relevant fs selection bit is set). I was thinking we could punt that behavioural change to another PR?

Reply:
I'd recommend just using skrifa for this. In particular, it handles things like metrics variations using the MVAR table and hides a lot of the nastiness. Take a look at https://docs.rs/skrifa/latest/skrifa/metrics/struct.Metrics.html

Reply:
So much better!! Used Skrifa in https://github.com/linebender/parley/pull/400/commits/06a96504b1a52b1151e22d3cd30fd96f6fbb5be5

Reply:
FYI: Skrifa calculates strikeout thickness differently to swash.

Swash uses `post.underline_size()`

https://github.com/dfrg/swash/blob/356c725e8e5800a0fecceddc047cbaacd73cb622/src/metrics.rs#L185

Skrifa uses the os2 table:

https://github.com/googlefonts/fontations/blob/830f97905764e0c285c39e6360819770ab19ec80/skrifa/src/metrics.rs#L168

The difference has meant that some snapshots needed to be updated.

Reply:
When in doubt, it's usually safe to assume that skrifa is more correct :)